distributed/shuffle/_scheduler_plugin.py

```diff
@@ -96,10 +97,44 @@ async def barrier(self, id: ShuffleId, run_id: int, consistent: bool) -> None:
                 stimulus_id=f"p2p-barrier-inconsistent-{time()}",
             )
         msg = {"op": "shuffle_inputs_done", "shuffle_id": id, "run_id": run_id}
-        await self.scheduler.broadcast(
-            msg=msg,
-            workers=list(shuffle.participating_workers),
-        )
+        workers = list(shuffle.participating_workers)
+        no_progress = 0
+        while workers:
+            res = await self.scheduler.broadcast(
+                msg=msg,
+                workers=workers,
+                on_error="return",
+            )
+            before = len(workers)
+            workers = []
+            for w, r in res.items():
+                if r is None:
+                    continue
+                if isinstance(r, OSError):
+                    workers.append(w)
+                else:
+                    raise P2PConsistencyError(
+                        "Unexpected error encountered during barrier"
+                    )
+            workers = [w for w, r in res.items() if r is not None]
+            if workers:
+                logger.warning(
+                    "Failure during broadcast of %s, retrying.",
+                    shuffle.id,
+                )
+                if any(w not in self.scheduler.workers for w in workers):
+                    raise P2PConsistencyError(
+                        f"Worker {workers} left during shuffle {shuffle}"
+                    )
+                await asyncio.sleep(0.1)
+                if len(workers) == before:
+                    no_progress += 1
+                    if no_progress >= 3:
+                        raise P2PConsistencyError(
```

Comment:
I would raise a different type of error here. The P2P run is still internally consistent, it just failed.

Reply:
are you fine with a generic RuntimeError?

Reply:
Yes, that would work. For context, I've imagined the `P2PConsistencyError` to be exclusively used for state/runtime inconsistencies. For transfer and unpack tasks, these actually cause us to reschedule the task instead of erring. (Come to think of it, I'm wondering if we should reschedule when encountering these in the barrier as well.)

Reply:
> (Come to think of it, I'm wondering if we should reschedule when encountering these in the barrier as well.)

I think the retry should be sufficient for now. Either way, this is out of scope for this PR

Reply:
I just checked and realized that we do _not_ reschedule upon `P2PConsistencyError`s but raise hard. We only reschedule on `ShuffleClosedError`.